client/include/Maszyna.Diagnostics/Trace.hxx

```diff
@@ -0,0 +1,182 @@
+            switch (value)
+            {
+            case TraceLevel::Disabled:
+                return "Disabled";
+            case TraceLevel::Debug:
+                return "Debug";
+            case TraceLevel::Normal:
+                return "Normal";
+            case TraceLevel::Warning:
+                return "Warning";
+            case TraceLevel::Error:
+                return "Error";
+            case TraceLevel::Critical:
+                return "Critical";
+            default:
+                return "Unknown";
+            }
+        }
+    }
+
+    class Trace final
+    {
+    public:
+        static bool Initialize(const char* filename);
+        static bool Shutdown();
+
+    public:
+        static void Fail();
+        static void Fail(const char* message);
+        static void Fail(const char* message, const char* details);
+        // Unconditionally dispatches message.
+        static bool Dispatch(TraceLevel::Enum level, const char* category, const char* format, ...);
+        // Called by macros, to check whether need to emit Dispatch call (along with built call frame).
+        static bool CanDispatch(TraceLevel::Enum level);
+        static void SetLevel(TraceLevel::Enum level);
+
+    private:
+        static TraceLevel::Enum _trace_level;
+    };
+
+    inline bool Trace::CanDispatch(TraceLevel::Enum level)
+    {
+        return static_cast<uintptr_t>(level) < static_cast<uintptr_t>(_trace_level);
+    }
+
+    inline void Trace::SetLevel(TraceLevel::Enum level)
+    {
+        _trace_level = level;
+    }
+
+    // Base for trace switches.
+    struct TraceSwitchBase
+    {
+    private:
+        const char* _name;
+        TraceLevel::Enum _level;
+        TraceLevel::Enum _compileTimeLevel;
+
+    public:
+        TraceSwitchBase(const char* name, TraceLevel::Enum defaultLevel, TraceLevel::Enum compileTimeLevel)
+            : _name(name)
+            , _level(defaultLevel)
+            , _compileTimeLevel(compileTimeLevel)
+        {
+        }
+
+    public:
+        inline const char* GetName() const
+        {
+            return _name;
+        }
+
+        inline bool CanDispatch(TraceLevel::Enum level) const
+        {
+            return level <= _level;
+        }
+
+        inline void SetLevel(TraceLevel::Enum level)
+        {
+            _level = level;
+        }
+    };
+
+    // Trace switch implementation. Default level and compile time levels
+    template <TraceLevel::Enum VDefaultLevel, TraceLevel::Enum VCompileTimeLevel>
+    struct TraceSwitch : public TraceSwitchBase
+    {
+    public:
+        enum
+        {
+            DefaultLevel = VDefaultLevel,
+            CompileTimeLevel = VCompileTimeLevel
+        };
+
+    public:
+        inline TraceSwitch(const char* name)
+            : TraceSwitchBase(name, VDefaultLevel, VCompileTimeLevel)
+        {
+        }
+    };
+}
+}
+
+
+#define __MASZYNA_DECLARE_TRACE_SWITCH_EXTERN(SwitchName, DefaultLevel, CompileTimeLevel) \
```

Comment:
Do we really need that macro? If you make CanDispatch constexpr the level > CompileTimeLevel should get optimized out.